cvprac/cvp_api.py

```diff
@@ -3207,3 +3207,360 @@ def workspace_build_status(self, workspace_id, build_id):
             if self.clnt.apiversion >= 6.0:
                 return self.clnt.get(url, timeout=self.request_timeout)
         return self.clnt.get(url, timeout=self.request_timeout)
+
+
+    def get_change_control(self, cc_id, cc_time=None):
+        ''' Get the configuration and status of a change control using Resource APIs.
+
+            Args:
+               cc_id (str): The ID of the change control.
+               time (str): Time indicates the time for which you are interested in the data.
+                    If no time is given, the server will use the time at which it makes the request.
+            Returns:
+               response (dict): A dict that contains...
+                  Ex: {"value":{"key":{"id":"rL6Tog6UU"}, "change":{"name":"Change 20211213_210554",
+                       "rootStageId":"kZUWqyIArD", "stages":{"values":{"kZUWqyIArD":{"name":"Change 20211213_210554 Root",
+                       "rows":{"values":[{"values":["vazWhKyVRR"]}]}}, "vazWhKyVRR":{"name":"Update Config",
+                       "action":{"name":"task", "timeout":3000, "args":{"values":{"TaskID":"538"}}}}}},
+                       "notes":"", "time":"2021-12-13T21:05:58.813750128Z", "user":"cvpadmin"},
+                       "approve":{"value":true, "time":"2021-12-13T21:11:26.788753264Z",
+                       "user":"cvpadmin"}}, "time":"2021-12-13T21:11:26.788753264Z"}%
+        '''
+        if cc_time is None:
+            params = 'key.id={}'.format(cc_id)
+        else:
+            params = 'key.id={}&time={}'.format(cc_id, cc_time)
+        url = '/api/resources/changecontrol/v1/ChangeControl?' + params
+        # For on-prem check the version as it is only supported from 2021.2.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion >= 6.0:
+                return self.clnt.get(url, timeout=self.request_timeout)
+        return self.clnt.get(url, timeout=self.request_timeout)
+
+
+    def get_all_change_controls(self):
+        ''' Get the configuration and status of all Change Controls using Resource APIs.
+
+            Returns:
+               response (dict): A dict that contains a list of all Change Controls.
+        '''
+        url = '/api/resources/changecontrol/v1/ChangeControl/all'
+        # For on-prem check the version as it is only supported from 2021.2.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion >= 6.0:
+                self.log.debug('v6 {}'.format(url))
+                response = self.clnt.get(url, timeout=self.request_timeout)
+                return response
+        return self.clnt.get(url, timeout=self.request_timeout)
+
+
+    def get_change_control_approval(self, cc_id, cc_time=None):
+        ''' Get the state of a specific Change Control's approve config using Resource APIs.
+
```

Comment:
yes, thanks for raising this, this is expected behaviour with the v1 API,  I'll add some error handling to this, basically today if a CC was never approved you cannot get approval status because the key doesn't exist in the db yet (will make this clear in the cloudvision-apis examples doc too)

Reply:
added error handling for this now